File: java/src/priorityQueue/MergingTables.java
package priorityQueue;

import java.util.Arrays;
import java.util.Scanner;

public class MergingTables {
    int n;
    int m;
    int[] P;
    int[] R;
    int[][] O;
    int maxSize;

    public MergingTables(int n, int m, int[] P, int[] R, int[][] O, int maxSize) {
        this.n = n;
        this.m = m;
        this.P = P;
        this.R = R;
        this.O = O;
        this.maxSize = maxSize;
    }

    private int find(int i) {
        if (i != this.P[i])
            this.P[i] = find(this.P[i]);
        return this.P[i];
    }

    private int union(int d, int s) {
        int d_id = find(d);
        int s_id = find(s);

        if (d_id != s_id) {
            this.R[d_id] += this.R[s_id];
            this.R[s_id] = 0;
            P[s_id] = d_id;
        }

        return R[d_id];
    }

    public void run() {
        int size = 0;

        for (int[] item : this.O) {
            size = union(item[0] - 1, item[1] - 1);

            if (size > this.maxSize)
                this.maxSize = size;

            System.out.println(this.maxSize);
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();

        int[] P = new int[n];
        int[] R = new int[n];
        int maxSize = 0;
        for (int i = 0;  i < n; i++) {
            P[i] = i;
            R[i] = scanner.nextInt();
            if (R[i] > maxSize)
                maxSize = R[i];
        }

        int[][] O = new int[m][2];
        for (int i = 0; i < m; i++) {
            O[i][0] = scanner.nextInt();
            O[i][1] = scanner.nextInt();
        }

        MergingTables merge = new MergingTables(n, m, P, R, O, maxSize);
        merge.run();
    }
}

File: python/priorityQueue/mergingTables.py

from random import randint

def find(i, P):
    if i != P[i]:
        P[i] = find(P[i], P)
    return P[i]

def union(d, s, P, R):
    d_id = find(d, P)
    s_id = find(s, P)

    if d_id != s_id:
        R[d_id] += R[s_id]
        R[s_id] = 0
        P[s_id] = d_id

    return R[d_id]

def merge(n, m, O, P, R):
    maxSize = max(R)

    for item in O:
        size = union(item[0] - 1, item[1] - 1, P, R)
        
        if size > maxSize:
            maxSize = size
        print(maxSize, end = ' ')

# input
n, m = map(int, input().split())
R = [*map(int, input().split())]
O = []
exec("O.append([*map(int, input().split())]);" * m)
P = [i for i in range(n)]

merge(n, m, O, P, R)


# random check
'''
N, M = map(int, input().split())

while True:
    n = randint(1, N)
    m = randint(1, M)

    R = [None] * n
    P = [None] * n
    for i in range(n):
        R[i] = randint(0, 10000)
        P[i] = i

    O = [None] * m
    for i in range(m):
        O[i] = [randint(1, n), randint(1, n)]

    merge(n, m, O, P, R)
'''
